plugins/kafka/kafka.ecllib

```diff
@@ -0,0 +1,255 @@
+        multiNodeDS := NORMALIZE
+            (
+                messageContainer,
+                LEFT.messages,
+                TRANSFORM
+                    (
+                        KafkaMessage,
+                        SELF := RIGHT
+                    ),
+                LOCAL
+            );
+
+        resultDS := MAP
+            (
+                numberOfPartitions = 0                      =>  DATASET([], KafkaMessage),
+                numberOfPartitions = 1 AND CLUSTERSIZE = 1  =>  singleNodeDS,
+                multiNodeDS
+            );
+
+        RETURN resultDS;
+
+    END;
+
+    /**
+     * Given a set of messages, presumably just consumed from an Apache Kafka
+     * cluster, summarize the last-read message offsets on a per-partition basis.
+     * This is useful for logging/saving the last messages read during a
+     * particular run, which can then be used to restore system state if you
+     * have to re-consume older messages (see SetMessageOffsets() function).
+     *
+     * @param   messages    A dataset of consumed messages; REQUIRED
+     *
+     * @return  A new dataset containing a summary of partitions and their
+     *          associated last-read message offsets.
+     */
+    EXPORT DATASET(KafkaMessageOffset) LastMessageOffsets(DATASET(KafkaMessage) messages) := FUNCTION
+        t := TABLE
+            (
+                messages,
+                {
+                    partitionNum,
+                    INTEGER8    offset := MAX(GROUP, offset)
+                },
+                partitionNum,
+                MERGE
+            );
+
+        f := PROJECT(t, TRANSFORM(KafkaMessageOffset, SELF := LEFT), LOCAL);
+
+        RETURN f;
+    END;
+
+    /**
+     * Resets the last-read partition offsets to the values in the given dataset.
+     * This is useful for "rewinding" message reading to an earlier point.  The
+     * next call to GetMessages() will start consuming at the points described
+     * in the dataset.
+     *
+     * @param   offsets     A dataset of of partitions and the offsets to which
+     *                      you want to set each, like the result from a call
+     *                      to LastMessageOffsets(); REQUIRED
+     *
+     * @return  The number of partitions set
+     */
+    EXPORT UNSIGNED4 SetMessageOffsets(DATASET(KafkaMessageOffset) offsets) := FUNCTION
+
+        // Distribute the offset data so that each partition lines up on the right node
+        distOffsets := DISTRIBUTE(offsets, partitionNum);
+
+        // Set the offset for each partition on each node
+        result := PROJECT
+            (
+                distOffsets,
+                TRANSFORM
+                    (
+                        KafkaMessageOffset,
+                        SELF.offset := kafka.SetMessageOffset(brokers, topic, consumerGroup, LEFT.partitionNum, LEFT.offset),
+                        SELF := LEFT
+                    ),
+                LOCAL
```

Comment:
trivial: LOCAL has no effect (It only has an affect on PROJECT if COUNTER is used)